module/state_synchronization/execution_data_service.go

```diff
@@ -2,430 +2,44 @@ package state_synchronization
-	// could fill the blob tree with lots of unnecessary data by appending it at the end of the serialized data for each level.
-	// It's possible that we could detect this and fail deserialization using something like the following:
-	// https://github.com/onflow/flow-go/blob/bd5320719266b045ae2cac954f6a56e1e79560eb/engine/access/rest/handlers.go#L189-L193
-	// Eventually, we will need to implement validation logic on Verification nodes to slash this behavior.
-
-	return v, bcr.TotalBytesRead(), nil
-}
-
-// Get gets the ExecutionData for the given root CID from the blobservice.
-// The returned error will be:
-// - MalformedDataError if some level of the blob tree cannot be properly deserialized
-// - BlobSizeLimitExceededError if any blob in the blob tree exceeds the maximum blob size
-// - BlobNotFoundError if some CID in the blob tree could not be found from the blobservice
-func (s *executionDataServiceImpl) Get(ctx context.Context, rootID flow.Identifier) (*ExecutionData, error) {
-	rootCid := flow.IdToCid(rootID)
-
-	logger := s.logger.With().Str("cid", rootCid.String()).Logger()
-	logger.Debug().Msg("getting execution data")
-
-	s.metrics.ExecutionDataGetStarted()
-
-	start := time.Now()
-	cids := []cid.Cid{rootCid}
-
-	var blobTreeSize uint64
-
-	for i := uint(0); i <= defaultMaxBlobTreeDepth; i++ {
-		v, totalBytes, err := s.getBlobs(ctx, cids, logger)
-
-		if err != nil {
-			s.metrics.ExecutionDataGetFinished(time.Since(start), false, blobTreeSize)
-
-			return nil, fmt.Errorf("failed to get level %d of blob tree: %w", i, err)
-		}
-
-		blobTreeSize += totalBytes
-
-		switch v := v.(type) {
-		case *ExecutionData:
-			s.metrics.ExecutionDataGetFinished(time.Since(start), true, blobTreeSize)
-
-			return v, nil
-		case *[]cid.Cid:
-			cids = *v
-		}
-	}
-
-	return nil, ErrBlobTreeDepthExceeded
-}
-
-// MalformedDataError is returned when malformed data is found at some level of the requested
-// blob tree. It likely indicates that the tree was generated incorrectly, and hence the request
-// should not be retried.
-type MalformedDataError struct {
-	err error
-}
-
-func (e *MalformedDataError) Error() string {
-	return fmt.Sprintf("malformed data: %v", e.err)
-}
-
-func (e *MalformedDataError) Unwrap() error { return e.err }
-
-// BlobSizeLimitExceededError is returned when a blob exceeds the maximum size allowed.
-type BlobSizeLimitExceededError struct {
-	cid cid.Cid
-}
-
-func (e *BlobSizeLimitExceededError) Error() string {
-	return fmt.Sprintf("blob %v exceeds maximum blob size", e.cid.String())
-}
-
-// BlobNotFoundError is returned when the blobservice failed to find a blob.
-type BlobNotFoundError struct {
-	cid cid.Cid
-}
-
-func (e *BlobNotFoundError) Error() string {
-	return fmt.Sprintf("blob %v not found", e.cid.String())
+const DefaultMaxBlobSize = 1 << 20 // 1MiB
+
+// ExecutionDataAdder handles adding execution data to a blobservice
```

Comment:
I think some extended documentation here would help with understanding the code, eg:
* Adder corresponds to data producer/provider (execution node)
* Adding data to the blob service stores it on the node's disk for future requests, and makes it available to other nodes via bitswap

Likewise for the Getter interface